File: src/test/fixtures/kotlin/CognitiveRules.kt
// straightLine: cognitive score 0.
fun straightLine(ready: Boolean, enabled: Boolean, urgent: Boolean): Boolean {
    return ready
}

// booleanRun: cognitive score 1.
fun booleanRun(ready: Boolean, enabled: Boolean, urgent: Boolean): Boolean {
    return ready && enabled && urgent
}

// mixedRuns: cognitive score 3.
fun mixedRuns(ready: Boolean, enabled: Boolean, urgent: Boolean): Boolean {
    return ready && enabled || urgent && ready
}

// groupedRun: cognitive score 1.
fun groupedRun(ready: Boolean, enabled: Boolean, urgent: Boolean): Boolean {
    return ready && (enabled && urgent)
}

// negatedGroup: cognitive score 2.
fun negatedGroup(ready: Boolean, enabled: Boolean, urgent: Boolean): Boolean {
    return ready && !(enabled && urgent)
}

// plainElse: cognitive score 2.
fun plainElse(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    if (ready) { deliver() } else { queue() }
}

// flatChain: cognitive score 3.
fun flatChain(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    if (ready) { deliver() } else if (urgent) { expedite() } else { queue() }
}

// nestedChain: cognitive score 4.
fun nestedChain(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    if (enabled) { if (ready) { deliver() } else if (urgent) { expedite() } }
}

// nestedChecks: cognitive score 3.
fun nestedChecks(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    if (enabled) { if (ready) { deliver() } }
}

// dispatch: cognitive score 3.
fun dispatch(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    when (ready) { true -> if (enabled) deliver(); false -> queue() }
}

// loopBody: cognitive score 3.
fun loopBody(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    while (ready) { if (enabled) deliver() }
}

// lambdaBody: cognitive score 2.
fun lambdaBody(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    val action = { if (ready) deliver() }; action()
}

// emptyLambda: cognitive score 0.
fun emptyLambda(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    val action = { deliver() }; action()
}

// guardReturn: cognitive score 1.
fun guardReturn(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    if (!ready) return
    deliver()
}

// unbracedChecks: cognitive score 3.
fun unbracedChecks(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    if (enabled) if (ready) deliver();
}

// doLoop: cognitive score 3.
fun doLoop(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    do { if (ready) deliver(); } while (enabled);
}

// nestedFunction: cognitive score 2.
fun nestedFunction(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    audit(); fun action() { if (ready) deliver() }; action()
}

// emptyFunction: cognitive score 0.
fun emptyFunction(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    audit(); fun action() { deliver() }; action()
}

// cleanup: cognitive score 0.
fun cleanup(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    try { deliver() } finally { cleanup() }
}

// catches: cognitive score 4.
fun catches(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    try { deliver() } catch (error: TimeoutException) { if (ready) queue() } catch (error: IllegalArgumentException) { reject() }
}

// labelJump: cognitive score 2.
fun labelJump(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    dispatch@ while (ready) { break@dispatch }
}

// labelContinue: cognitive score 2.
fun labelContinue(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    dispatch@ while (ready) { continue@dispatch }
}

// labelReturn: cognitive score 1.
fun labelReturn(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    dispatch@ while (ready) { return@dispatch }
}

// ordinaryBreak: cognitive score 1.
fun ordinaryBreak(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    while (ready) { break }
}

// bracedElseCheck: cognitive score 4.
fun bracedElseCheck(ready: Boolean, enabled: Boolean, urgent: Boolean): Unit {
    if (enabled) { deliver(); } else { if (ready) { deliver(); } }
}

File: src/test/fixtures/typescript/cognitiveRules.ts
// straightLine: cognitive score 0.
function straightLine(ready: boolean, enabled: boolean, urgent: boolean) {
    return ready;
}

// booleanRun: cognitive score 1.
function booleanRun(ready: boolean, enabled: boolean, urgent: boolean) {
    return ready && enabled && urgent;
}

// mixedRuns: cognitive score 3.
function mixedRuns(ready: boolean, enabled: boolean, urgent: boolean) {
    return ready && enabled || urgent && ready;
}

// groupedRun: cognitive score 1.
function groupedRun(ready: boolean, enabled: boolean, urgent: boolean) {
    return ready && (enabled && urgent);
}

// negatedGroup: cognitive score 2.
function negatedGroup(ready: boolean, enabled: boolean, urgent: boolean) {
    return ready && !(enabled && urgent);
}

// plainElse: cognitive score 2.
function plainElse(ready: boolean, enabled: boolean, urgent: boolean) {
    if (ready) { deliver(); } else { queue(); }
}

// flatChain: cognitive score 3.
function flatChain(ready: boolean, enabled: boolean, urgent: boolean) {
    if (ready) { deliver(); } else if (urgent) { expedite(); } else { queue(); }
}

// nestedChain: cognitive score 4.
function nestedChain(ready: boolean, enabled: boolean, urgent: boolean) {
    if (enabled) { if (ready) { deliver(); } else if (urgent) { expedite(); } }
}

// nestedChecks: cognitive score 3.
function nestedChecks(ready: boolean, enabled: boolean, urgent: boolean) {
    if (enabled) { if (ready) { deliver(); } }
}

// dispatch: cognitive score 3.
function dispatch(ready: boolean, enabled: boolean, urgent: boolean) {
    switch (ready) { case true: if (enabled) deliver(); break; default: queue(); }
}

// loopBody: cognitive score 3.
function loopBody(ready: boolean, enabled: boolean, urgent: boolean) {
    while (ready) { if (enabled) deliver(); }
}

// lambdaBody: cognitive score 2.
function lambdaBody(ready: boolean, enabled: boolean, urgent: boolean) {
    const action = () => { if (ready) deliver(); }; action();
}

// emptyLambda: cognitive score 0.
function emptyLambda(ready: boolean, enabled: boolean, urgent: boolean) {
    const action = () => deliver(); action();
}

// guardReturn: cognitive score 1.
function guardReturn(ready: boolean, enabled: boolean, urgent: boolean) {
    if (!ready) return; deliver();
}

// unbracedChecks: cognitive score 3.
function unbracedChecks(ready: boolean, enabled: boolean, urgent: boolean) {
    if (enabled) if (ready) deliver();
}

// doLoop: cognitive score 3.
function doLoop(ready: boolean, enabled: boolean, urgent: boolean) {
    do { if (ready) deliver(); } while (enabled);
}

// nestedFunction: cognitive score 2.
function nestedFunction(ready: boolean, enabled: boolean, urgent: boolean) {
    audit(); function action() { if (ready) deliver(); } action();
}

// emptyFunction: cognitive score 0.
function emptyFunction(ready: boolean, enabled: boolean, urgent: boolean) {
    audit(); function action() { deliver(); } action();
}

// cleanup: cognitive score 0.
function cleanup(ready: boolean, enabled: boolean, urgent: boolean) {
    try { deliver(); } finally { cleanup(); }
}

// catches: cognitive score 3.
function catches(ready: boolean, enabled: boolean, urgent: boolean) {
    try { deliver(); } catch (error) { if (ready) queue(); }
}

// labelJump: cognitive score 2.
function labelJump(ready: boolean, enabled: boolean, urgent: boolean) {
    dispatch: while (ready) { break dispatch; }
}

// ordinaryBreak: cognitive score 1.
function ordinaryBreak(ready: boolean, enabled: boolean, urgent: boolean) {
    while (ready) { break; }
}

// bracedElseCheck: cognitive score 4.
function bracedElseCheck(ready: boolean, enabled: boolean, urgent: boolean) {
    if (enabled) { deliver(); } else { if (ready) { deliver(); } }
}

// multilineRuns: cognitive score 2.
function multilineRuns(ready: boolean, enabled: boolean, urgent: boolean) {
    return ready
        && enabled
        || urgent;
}

// conditionalCondition: cognitive score 2.
function conditionalCondition(ready: boolean, enabled: boolean, urgent: boolean) {
    if (ready && enabled) deliver();
}

// booleanArgument: cognitive score 1.
function booleanArgument(ready: boolean, enabled: boolean, urgent: boolean) {
    record(ready && enabled);
}

// booleanAssignment: cognitive score 1.
function booleanAssignment(ready: boolean, enabled: boolean, urgent: boolean) {
    const allowed = ready && enabled; record(allowed);
}
